addons/stock/stock.py

```diff
@@ -2440,12 +2440,14 @@ def action_cancel(self, cr, uid, ids, context=None):
                     procurement_ids = procurement_obj.search(cr, uid, [('move_dest_id', '=', move.id)], context=context)
                     procurement_obj.cancel(cr, uid, procurement_ids, context=context)
             else:
-                if move.move_dest_id:
+                if move.move_dest_id and not move.origin_returned_move_id:
                     if move.propagate:
                         self.action_cancel(cr, uid, [move.move_dest_id.id], context=context)
                     elif move.move_dest_id.state == 'waiting':
                         #If waiting, the chain will be broken and we are not sure if we can still wait for it (=> could take from stock instead)
                         self.write(cr, uid, [move.move_dest_id.id], {'state': 'confirmed'}, context=context)
+                elif move.origin_returned_move_id and move.origin_returned_move_id.move_dest_id:
+                    self.action_assign(cr, uid, [move.origin_returned_move_id.move_dest_id.id], context=context)
```

Comment:
I think what you wanted to do is move.origin_returned_move_id.origin_returned_move_id.move_dest_id